### PR TITLE
Of consumption select batch and warehouse

### PR DESCRIPTION
**Pull request by atm-lena**
# NEW : Of consumption select batch and warehouse

**Where ?** OF Card, action "consumeorproduce" or "consumeandproduceall", consumption part

1. Add stock information in select batch

1. Add automatic behaviors between selects (batch and warehouse) : 

- If no warehouse is specified, the serial number list displays all serial numbers of the product and the total physical inventory of the product
- If no warehouse is filled in, but I select a serial number, if this serial number only has stock in one warehouse, the warehouse is filled in automatically
- If a warehouse is specified, only the serial numbers with stock in this warehouse appear in the drop-down list (the stock of the serial number indicated is then only the stock in the warehouse specified)

### CODE REVIEW

File: htdocs/mrp/ajax/interface.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+global $db, $langs;
+
+require '../../main.inc.php'; // Load $user and permissions
+
+$warehouse_id = GETPOST('warehouse_id', 'int');
+$batch_id = GETPOST('batch_id', 'int');
+$fk_product = GETPOST('product_id', 'int');
+$action = GETPOST('action', 'alphanohtml');
+
+if ($action == 'updateselectbatchbywarehouse') {
```
Comment (eldy) [inline htdocs/mrp/ajax/interface.php:35]: Missing a test on a permission

File: htdocs/mrp/ajax/interface.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	$resql = $db->query($sql);
+
+	if ($resql) {
+		while ($obj = $db->fetch_object($resql)) {
+			$TRes[$obj->batch] += $obj->qty;
+		}
+	}
+
+	print json_encode($TRes);
+} elseif ($action == 'updateselectwarehousebybatch') {
```
Comment (eldy) [inline htdocs/mrp/ajax/interface.php:55]: Missing a test on a permission

File: htdocs/mrp/ajax/interface.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+$permissiontoproduce = GETPOST('permissiontoproduce', 'int');
+
+if ($action == 'updateselectbatchbywarehouse' && $permissiontoproduce) {
+	$TRes = array();
+
+	$sql = "SELECT pb.batch, pb.rowid, ps.fk_entrepot, pb.qty, e.ref as label, ps.fk_product";
+	$sql .= " FROM " . MAIN_DB_PREFIX . "product_batch as pb";
+	$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "product_stock as ps on ps.rowid = pb.fk_product_stock";
+	$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "entrepot as e on e.rowid = ps.fk_entrepot AND e.entity IN (" . getEntity('stock') . ")";
+	$sql .= " WHERE ps.fk_product = '" . $fk_product . "'";
```
Comment (eldy) [inline htdocs/mrp/ajax/interface.php:43]: Must be 
$sql .= " WHERE ps.fk_product = " . ((int) $fk_product);
Not quote on a comparison with a field that is integer (pb with not optimized sql execution plan due to implicit cast done by database engine).

**Comment by atm-lena** [comment]
@eldy ?

File: htdocs/product/class/html.formproduct.class.php [hunk 4]
```diff
@@ -655,13 +655,11 @@ public function selectLotDataList($htmlname = 'batch_id', $empty = 0, $fk_produc
 		}
 
 		foreach ($productIdArray as $productId) {
-			if (array_key_exists($productId, $this->cache_lot)) {
```
Comment (eldy) [inline htdocs/product/class/html.formproduct.class.php:658]: Is removing this line to avoid warning was done on purpose ?

Comment (atm-lena) [reply htdocs/product/class/html.formproduct.class.php:658]: I seems like a mistake, I fix it

File: htdocs/mrp/ajax/interface.php [hunk 5]
```diff
@@ -61,7 +61,7 @@
 	$sql .= " JOIN " . MAIN_DB_PREFIX . "product_stock as ps on ps.rowid = pb.fk_product_stock";
 	$sql .= " JOIN " . MAIN_DB_PREFIX . "entrepot as e on e.rowid = ps.fk_entrepot AND e.entity IN (" . getEntity('stock') . ")";
 	$sql .= " WHERE ps.fk_product = " .((int) $fk_product);
-	if ($batch_id > 0) $sql.= " AND pb.batch = '" . ((int) $batch_id) . "'";
+	if ($batch) $sql.= " AND pb.batch = '" . $batch . "'";
```
Comment (eldy) [inline htdocs/mrp/ajax/interface.php:64]: Missing $db->escape($batch)

File: htdocs/product/class/html.formproduct.class.php (hunk 4 above)
Comment (eldy) [reply htdocs/product/class/html.formproduct.class.php:658]: I still don't see the protection in PR.

File: htdocs/product/class/html.formproduct.class.php [hunk 6]
```diff
@@ -655,13 +655,11 @@ public function selectLotDataList($htmlname = 'batch_id', $empty = 0, $fk_produc
...
-				foreach ($this->cache_lot[$productId] as $id => $arraytypes) {
-					if (empty($fk_entrepot) || $fk_entrepot == $arraytypes['entrepot_id']) {
-						$label = $arraytypes['entrepot_label'] . ' - ';
-						$label .= $arraytypes['batch'];
-						$out .= '<option>' . $arraytypes['batch'] . '</option>';
-					}
+			foreach ($this->cache_lot[$productId] as $id => $arraytypes) {
```
Comment (eldy) [inline htdocs/product/class/html.formproduct.class.php:658]: The test if (array_key_exists) required for php8 compatibility seems missing. Can you restore it ?

Comment (atm-lena) [reply htdocs/product/class/html.formproduct.class.php]: @eldy Done

**Comment by atm-lena** [comment]
@eldy

File: htdocs/mrp/ajax/interface.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	define('NOREQUIREHTML', '1');
+}
+if (!defined('NOREQUIREAJAX')) {
+	define('NOREQUIREAJAX', '1');
+}
+
+global $db, $langs;
+
+require '../../main.inc.php'; // Load $user and permissions
+
```
Comment (eldy) [inline htdocs/mrp/ajax/interface.php:29]: Can you add here this:

top_httphead("text/json");
I let you see the content of this metho to understand why...

File: htdocs/product/class/html.formproduct.class.php [hunk 8]
```diff
@@ -660,7 +660,7 @@ public function selectLotDataList($htmlname = 'batch_id', $empty = 0, $fk_produc
 					if (empty($fk_entrepot) || $fk_entrepot == $arraytypes['entrepot_id']) {
 						$label = $arraytypes['entrepot_label'] . ' - ';
 						$label .= $arraytypes['batch'];
-						$out .= '<option>' . $arraytypes['batch'] . '</option>';
+						$out .= '<option value="' . $arraytypes['batch'] . '">(' . $langs->trans('Stock Total') . ': ' . $arraytypes['qty'] . ')</option>';
```
Comment (eldy) [inline htdocs/product/class/html.formproduct.class.php:663]: The $arraytypes['batch'] is added into value but does not appear into visible part ?  (between <option> and </option>

Comment (atm-gauthier) [reply htdocs/product/class/html.formproduct.class.php:663]: Hi @eldy, I just tried what @atm-lena did, and it seems to work properly :

Before : 
image

After :
image

**Pull request merged**